smarts/core/smarts.py

```diff
@@ -790,6 +822,24 @@ def _reset_providers(self):
         for provider in self.providers:
             provider.reset()
 
+    def _handle_provider(self, provider: Provider, provider_error):
+        provider_problem = bool(provider_error or not provider.connected)
+        if not provider_problem:
+            return
+
+        recovery_flags = self._provider_recovery_flags.get(
+            provider, ProviderRecoveryFlags.EXPERIMENT_REQUIRED
+        )
+        recovered = False
+        if recovery_flags & ProviderRecoveryFlags.ATTEMPT_RECOVERY:
+            recovered = provider.recover(self._scenario, self.elapsed_sim_time)
+
+        if not recovered:
+            if recovery_flags & ProviderRecoveryFlags.EPISODE_REQUIRED:
+                self._reset_required = True
+            elif recovery_flags & ProviderRecoveryFlags.EXPERIMENT_REQUIRED:
+                raise provider_error
```

Comment:
The meaning of required is that it will re-raise the error it had if it cannot (or will not recover) from the error.